sympy/solvers/ode.py

```diff
@@ -4003,6 +4012,42 @@ def _frobenius(n, m, p0, q0, p, q, x0, x, c, check=None):
 
     return frobdict
 
+def _check_substitution_type_match(eq, func):
+    r"""
+    Matches any differential equation that `order_reducing_substitution` can solve.
+    For this to work equation should have a minimum order of derivative to be 1, i.e.,
+    `f(x)` should not be present.
+    """
+    x = func.args[0]
+    f = func.func
+    order_to_subs = 0
+    D = Dummy()
+    for nsubs_try in range(1,ode_order(eq,f(x))):
+        if eq.subs(f(x).diff(x, nsubs_try), D).has(f(x)):
+            break
+        else:
+            order_to_subs += 1
+    if order_to_subs > 0:
+        return {'var':order_to_subs, 'solutions':True}
+    else:
+        return {'var':order_to_subs, 'solutions':False}
+
+# Use repeated substitution until we do not have a function independent of derivative
+def ode_order_reducing_substitution(eq, func, order, match):
+    r"""
+    Substitutes lowest order derivate in equation to function with order
+    of derivative as 0.Eg `f^(n)(x) = g(x)`, where n is the least order derivate.
+    match[var] or n here is how many times the function if solved is to be integrated
+    to  get f(x) since `g(x) = f^(match[var])(x)`.
+    """
+    x = func.args[0]
+    f = func.func
+    g = Function('g')
+    eq = eq.subs(f(x).diff(x, match['var']), g(x))
```

Comment:
`g(x)` might already be present. You will have to pick a function which does not already appear in the eq:

```python
# get unique function name
while 1:
    g = Function(Dummy().name)
    if not eq.has(g(x)):
        break
```

Reply:
Sure.But I 'll have to summarize all those points first.

Reply:
So,
1: Change 'g' to dummy.
2: I wanted to return None but I don't know how dictionaries react to if keys are not found.Initially I added 
```
if match:
   if match["order_ reducing_substitution"]:
    .....

```
but that was requested to be changed I guess.

Reply:
The problem with what you had before was that you had two checks and the second was unnecessary. You could have just done:
```python
if r is not None:
    match['order_reducing_substitution'] = r
```

Reply:
```
>>> d =dict(a=9,b=5)
>>> d
{'a': 9, 'b': 5}
>>> d["c"]
Traceback (most recent call last):
  File "<pyshell#8>", line 1, in <module>
    d["c"]
KeyError: 'c'
>>> 
```

Reply:
Ok, I have corrected that.What else should I do?

Reply:
Done that.Oh, so the order in all_hints is the preference order of hints.